package/samplers/grey_wolf_optimization/grey_wolf_optimization.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+
+from typing import Any
+
+import numpy as np
+import optuna
+from optuna.distributions import BaseDistribution
+from optuna.samplers import RandomSampler
+import optunahub
+
+
+class GreyWolfOptimizationSampler(optunahub.load_module("samplers/simple").SimpleBaseSampler):  # type: ignore
+    def __init__(
+        self,
+        search_space: dict[str, BaseDistribution] | None = None,
+        population_size: int = 10,
+        max_iter: int = 40,
+        num_leaders: int = 3,
+        seed: int = 0,
+    ) -> None:
+        self.search_space = search_space
+        self.population_size = population_size
+        self.max_iter = max_iter
+        self.num_leaders = min(max(1, num_leaders), self.population_size // 2)
+        self._rng = np.random.RandomState(seed)
+        self.dim = 0
+        self.leaders: np.ndarray = np.array([])  # Leaders (alpha, beta, gamma, ...) positions
+        self.wolves: np.ndarray = np.array([])  # Wolf positions
+        self.fitnesses: np.ndarray = np.full(population_size, np.inf)  # Fitness values
+        self._random_sampler = RandomSampler(seed=seed)
+        self.queue: list[dict[str, Any]] = []  # Queue to hold candidate positions
+
+    def _lazy_init(self, search_space: dict[str, BaseDistribution]) -> None:
+        self.dim = len(search_space)
+        self.lower_bound = np.array([dist.low for dist in search_space.values()])
+        self.upper_bound = np.array([dist.high for dist in search_space.values()])
```

Comment:
If users use `trial.suggest_categorical`, the following error will occur:

> AttributeError: 'CategoricalDistribution' object has no attribute 'low'

But we have some workarounds, like specifying the search space for `__init__` in this class. We can work on it in a follow-up PR.